File: packages/w2v/src/training/negative_sampling.rs
use super::{ObjectiveLoss, ObjectiveScratch, objective_step};
use crate::{
    config::{Real, Status},
    random::Rng,
    trainer::Trainer,
};

pub fn train(
    trainer: &Trainer,
    target_token: usize,
    learning_rate: Real,
    negative_rng: &mut Rng,
    hidden: &[Real],
    hidden_gradient: &mut [Real],
    observe: bool,
) -> Result<ObjectiveLoss, Status> {
    let mut output_snapshot = vec![0.0; hidden.len()];
    train_with_scratch(
        trainer,
        target_token,
        learning_rate,
        negative_rng,
        hidden,
        ObjectiveScratch {
            hidden_gradient,
            output_snapshot: &mut output_snapshot,
        },
        observe,
    )
}

pub(crate) fn train_with_scratch(
    trainer: &Trainer,
    target_token: usize,
    learning_rate: Real,
    negative_rng: &mut Rng,
    hidden: &[Real],
    scratch: ObjectiveScratch<'_>,
    observe: bool,
) -> Result<ObjectiveLoss, Status> {
    let dimension = trainer.model.embedding_dimension;
    let mut loss = ObjectiveLoss::default();
    for sample_index in 0..=trainer.config.negative_sample_count {
        let (sampled_token, label) = if sample_index == 0 {
            (target_token, 1.0)
        } else {
            let (mut sampled_token, random_value) = trainer.negative_sampler.draw(negative_rng);
            if sampled_token == 0 {
                sampled_token =
                    (random_value % (trainer.vocab.entries.len() as u64 - 1) + 1) as usize;
            }
            if sampled_token == target_token {
                continue;
            }
            (sampled_token, 0.0)
        };
        let start = sampled_token * dimension;
        let output_row = &trainer.model.output_embeddings[start..start + dimension];
        objective_step(
            hidden,
            scratch.hidden_gradient,
            output_row,
            scratch.output_snapshot,
            trainer.config.update_strategy,
            |score| {
                let prediction = trainer.sigmoid_table.lookup(score);
                if !prediction.is_finite() {
                    return Err(Status::InvalidState);
                }
                if observe {
                    let probability = if label == 1.0 {
                        prediction
                    } else {
                        1.0 - prediction
                    };
                    let value = -(probability as f64).max(f64::EPSILON).ln();
                    if !value.is_finite() {
                        return Err(Status::InvalidState);
                    }
                    loss.sum += value;
                    loss.count += 1;
                }
                Ok(Some((label - prediction) * learning_rate))
            },
        )?;
    }
    Ok(loss)
}

File: packages/w2v/reference/src/training/negative_sampling.c
#include "negative_sampler.h"
#include "sigmoid_table.h"
#include "training.h"

void negative_sampling_train(
    ModelStep *step,
    const real *hidden,
    real *hidden_gradient)
{
    Trainer *trainer = step->trainer;
    size_t dimension = trainer->model->embedding_dimension;
    for (size_t sample_index = 0;
         sample_index <= trainer->config.negative_sample_count;
         sample_index++)
    {
        size_t sampled_token;
        real label;

        if (sample_index == 0)
        {
            sampled_token = step->target_token;
            label = 1;
        }
        else
        {
            uint64_t random_value;
            sampled_token = negative_sampler_draw(
                &trainer->negative_sampler,
                &step->worker->negative_rng,
                &random_value);
            if (sampled_token == 0)
            {
                sampled_token =
                    random_value % (trainer->vocab->size - 1) + 1;
            }
            if (sampled_token == step->target_token)
            {
                continue;
            }
            label = 0;
        }

        _Atomic uint32_t *output_row =
            trainer->model->output_embeddings + sampled_token * dimension;
        real score = objective_score(hidden, output_row, dimension);
        real prediction = sigmoid_table_lookup(
            &trainer->sigmoid_table,
            score);
        real gradient_scale =
            (label - prediction) * step->learning_rate;

        objective_apply_update(
            hidden,
            hidden_gradient,
            output_row,
            dimension,
            gradient_scale);
    }
}
